### DISCUSSION TITLE
Integrating vkGetPastPresentationTimingGOOGLE: where to start?

### DISCUSSION THREAD

**Discussion by badicsalex** [gfx-rs/wgpu#3902] [Q&A]
Hi,

I'm writing an Augmented Reality application, targeting android phones. Our top priority is that latency (the time between the buffer update and the rendered frame actually reaching the displayport cable) has to be the absolute minimum. Theoretically it could be around 25ms (one and a half frames at 60FPS). Right now we have 45ms, because we are forced to use Fifo presentation mode (i.e. triple buffering instead of double), and the only way to have preciuse frame timing is to let `.get_current_texture` block. I cannot stress how important that ~20ms additional delay is: it makes the image look like it's being held with rubber bands.

What I'd like to do is use `Mailbox` presentation mode and the `vkGetPastPresentationTimingGOOGLE` extension, which would tell me exactly how early I was with the `present` call, and adjust my wait times and periods according to that information.

I looked into using `as_hal` and `get_raw` and friends in `wgpu`, and then using the function pointer exposed by `ash`, but unfortunately I cannot get hold of the SwapchainKHR object deep within `wgpu`.

Now as an absolute minimum, it would be great to just get the `ash` `Device` and `SwapchainKHR` from regular `wgpu` calls, but I could also integrate the timing info into `wgpu` as an optional extension function. I just don't really know where to start, and what are the policies for Pull requests, etc.

BR,
Alex

**Comment by victorvde** [comment]
There has been some work done on this by @cwfitzgerald in the context of https://github.com/gfx-rs/wgpu/issues/2869 .

**Comment by badicsalex** [comment]
Thanks. I've been wondering why there's a `Adapter::get_presentation_timestamp` but nothing else related; it's because work is ongoing.
Sorry for not finding the existing issue. I'll continue discussion in #2869
